astacus/coordinator/list.py

```diff
@@ -3,12 +3,48 @@
 See LICENSE for details
 
 """
-
 from astacus.common import ipc, magic
-from astacus.common.storage import MultiStorage
+from astacus.common.storage import JsonStorage, MultiStorage
+from collections import defaultdict
+from collections.abc import Iterator
+
+
+def compute_deduplicated_snapshot_file_stats(manifest: ipc.BackupManifest) -> tuple[int, int]:
+    """Compute stats over snapshot files as identified by their hex digest.
+
+    There may be duplicate hex digests within nodes for multiple copies of the same data chunks.
+    Duplicates' size are aggregated as they require physical disk space during restore.
+    On the other hand, hex digests can be safely deduplicated across nodes. The ocurrence with
+    the highest associated files' count and maximum file size across all nodes is kept.
+    """
+    global_file_counter = defaultdict(list)
+    global_file_sizes = defaultdict(list)
+    num_files = 0
+    total_size = 0
+    for snapshot_result in manifest.snapshot_results:
+        if not snapshot_result.state:
+            num_files += snapshot_result.files
+            total_size += snapshot_result.total_size
+        else:
+            node_hexdigest_counter: defaultdict[str, int] = defaultdict(lambda: 0)
+            node_hexdigest_sizes: defaultdict[str, list[int]] = defaultdict(list)
+            for snapshot_file in snapshot_result.state.files:
+                node_hexdigest_counter[snapshot_file.hexdigest] += 1
+                node_hexdigest_sizes[snapshot_file.hexdigest].append(snapshot_file.file_size)
+            for hexdigest, count in node_hexdigest_counter.items():
+                node_hexdigest_max_size = max(node_hexdigest_sizes[hexdigest])
+                global_file_sizes[hexdigest].append(node_hexdigest_max_size)
+                global_file_counter[hexdigest].append(count)
+    # This is not strictly correct as it takes the global max size and not the size associated with
```

Comment:
size is constant (hexdigest is hash of the file -> likelihood of having same hash map to files with different sizes is very, very slim). I'm not sure if it affects anything here as such but the comment reads wrong.

Reply:
I'm ashamed haha